python/hail/utils/misc.py

```diff
@@ -17,19 +20,157 @@ def wrap_to_list(s):
+            assert field in index_obj._fields
+            inds = index_obj[field]._indices
+            if inds == index_obj._global_indices:
+                return "'{}' [globals]".format(field)
+            elif inds == index_obj._row_indices:
+                return "'{}' [row]".format(field)
+            elif inds == index_obj._col_indices:  # Table will never get here
+                return "'{}' [col]".format(field)
+            else:
+                assert inds == index_obj._entry_indices
+                return "'{}' [entry]".format(field)
+
+        s = ["{} instance has no field, method, or property '{}'".format(class_name, item)]
+        if any([field_matches, method_matches, prop_matches]):
+            s.append('\n    Did you mean:')
+            if field_matches:
+                l = []
+                for f in field_matches:
+                    l.extend(dd[f])
+                word = plural('field', len(l))
+                s.append('\n        Data {}: {}'.format(word, ', '.join(fmt_field(f) for f in l)))
+            if method_matches:
+                word = plural('method', len(method_matches))
+                s.append('\n        {} {}: {}'.format(class_name, word,
+                                                      ', '.join("'{}'".format(m) for m in method_matches)))
+            if prop_matches:
+                word = plural('property', len(prop_matches), 'properties')
+                s.append('\n        {} {}: {}'.format(class_name, word,
+                                                      ', '.join("'{}'".format(p) for p in prop_matches)))
+        else:
+            s.append("\n    Hint: use 'describe()' to show the names of all data fields.")
+        return ''.join(s)
+
+
+def get_nice_field_error(obj, item):
+    from hail.api2 import MatrixTable, GroupedMatrixTable, Table, GroupedTable
+    from hail.expr import TStruct
+    if isinstance(obj, MatrixTable):
+        class_name = 'MatrixTable'
+        index_obj = obj
+    elif isinstance(obj, GroupedMatrixTable):
+        class_name = 'GroupedMatrixTable'
+        index_obj = obj._parent
+    elif isinstance(obj, Table):
+        class_name = 'Table'
+        index_obj = obj
+    elif isinstance(obj, GroupedTable):
+        class_name = 'GroupedTable'
+        index_obj = obj._parent
+    else:
+        raise NotImplementedError(obj)
+
+    fields = obj._fields.keys()
+    dd = defaultdict(lambda: [])
+    for f in fields:
+        dd[f.lower()].append(f)
+
+    def fmt_field(field):
+        assert field in index_obj._fields
+        field_expr = index_obj[field]
+        inds = field_expr._indices
+        if inds == index_obj._global_indices:
+            return "'{}' [global field]".format(field)
+        elif inds == index_obj._row_indices:
+            return "'{}' [row field]".format(field)
+        elif inds == index_obj._col_indices:  # Table will never get here
+            return "'{}' [col field]".format(field)
+        else:
```

Comment:
Can you refactor this at all to minimize code repetition with `get_nice_attr_error`?